src/Extensions/SiteTreeExtension.php

```diff
@@ -56,45 +56,47 @@ public function getSocialMetaConfig()
         return $siteConfig->getCurrentSocialMetaConfig();
     }
 
-    public function MetaTags(&$tagString)
+    /**
+     * @param array $tags
+     */
+    public function MetaComponents(array &$tags)
     {
-        $socialMetaTags = [];
-
-        // update title tag if set
-        if (preg_match('/<title>.*<\/title>/', $tagString)) {
-            if ($this->owner->getSocialMetaValue('Title')) {
-                $formattedTitleTag = HTML::createTag(
-                    'title',
-                    [],
-                    $this->owner->getSocialMetaValue('Title')
-                );
-                $tagString = preg_replace('/<title>.*<\/title>/', $formattedTitleTag, $tagString);
-            }
+        // update title tag
+        if ($this->owner->getSocialMetaValue('Title')) {
+            $tags['title'] = [
+                'tag' => 'title',
+                'content' => $this->owner->getSocialMetaValue('Title'),
+            ];
         }
 
         // update meta description
-        $tagString = preg_replace('/<meta name="description"[^>]+>\\n?/', '', $tagString);
         if ($this->owner->getSocialMetaValue('Description')) {
-            $socialMetaTags[] = HTML::createTag('meta', [
-                'name'      =>  'description',
-                'content'   =>  $this->owner->getSocialMetaValue('Description')
-            ]);
+            $tags['description'] = [
+                'attributes' => [
+                    'name' => 'description',
+                    'content' => $this->owner->getSocialMetaValue('Description'),
+                ],
+            ];
         }
 
         $metaAuthor = $this->owner->getSocialMetaValue('Author');
         if ($metaAuthor) {
-            $socialMetaTags[] = HTML::createTag('meta', [
-                'name'      =>  'author',
-                'content'   =>  $metaAuthor
-            ]);
+            $tags['author'] = [
+                'attributes' => [
+                    'name' => 'author',
+                    'content' => $metaAuthor,
+                ],
+            ];
         }
 
         $canonicalURL = $this->owner->getSocialMetaValue('CanonicalURL');
         if ($canonicalURL) {
-            $socialMetaTags[] = HTML::createTag('link', [
-                'rel'   =>  'canonical',
-                'href'  =>  $canonicalURL
-            ]);
+            $tags['canonical'] = [
+                'attributes' => [
+                    'rel' => 'canonical',
+                    'href' => $canonicalURL,
+                ],
+            ];
```

Comment:
I'm a little confused on this one, here is a sample canonical from my local `<meta rel="canonical" href="http://socialmetadata.test/about-us/" />` which seems to be correct. I may be missing something obvious though.

Reply:
In some further testing, should the module respect the canonical set by `VirtualPage`? As it stands it will override that value resulting in the canonical being the virtual page's url vs the source page's url.

Reply:
it should be `<link`, not `<meta`, see https://developers.google.com/search/docs/advanced/crawling/consolidate-duplicate-urls#rel-canonical-link-method. That's incorrect in `VirtualPage` as well.

This is actually a bit weird with the CMS releases:
The change in `VirtualPage` is included in branch 4.7 (https://github.com/silverstripe/silverstripe-cms/blob/4.7/code/Model/VirtualPage.php#L139) and 4.8 (https://github.com/silverstripe/silverstripe-cms/blob/4.8/code/Model/VirtualPage.php#L139), but it's not actually in the tagged version 4.8.0 (https://github.com/silverstripe/silverstripe-cms/blob/4.8.0/code/Model/VirtualPage.php#L143). It is in 4.7.2 and 4.9.0-rc1. Not sure why.

But anyway: Our extension is on `SiteTree`, which means that once the `VirtualPage` change is actually in the release, the code in `VirtualPage` runs after our `SiteTree` extension and overwrites our changes. That's the behaviour we want. Assuming that the changes in `VirtualPage` finally go into 4.9.0, we can set the requirements for this module to CMS ^4.9. right?